packages/ra-core/src/dataProvider/useDataProvider.ts

```diff
@@ -247,34 +257,154 @@ const doQuery = ({
             resourceState,
             dispatch,
         });
+    } else if (mutationMode === 'optimistic') {
+        return performOptimisticQuery({
+            type,
+            payload,
+            resource,
+            action,
+            rest,
+            onSuccess,
+            onFailure,
+            dataProvider,
+            dispatch,
+            logoutIfAccessDenied,
+            allArguments,
+        });
+    } else if (mutationMode === 'undoable' || undoable) {
+        return performUndoableQuery({
+            type,
+            payload,
+            resource,
+            action,
+            rest,
+            onSuccess,
+            onFailure,
+            dataProvider,
+            dispatch,
+            logoutIfAccessDenied,
+            allArguments,
+        });
+    } else {
+        return performQuery({
+            type,
+            payload,
+            resource,
+            action,
+            rest,
+            onSuccess,
+            onFailure,
+            dataProvider,
+            dispatch,
+            logoutIfAccessDenied,
+            allArguments,
+        });
     }
-    return undoable
-        ? performUndoableQuery({
-              type,
-              payload,
-              resource,
-              action,
-              rest,
-              onSuccess,
-              onFailure,
-              dataProvider,
-              dispatch,
-              logoutIfAccessDenied,
-              allArguments,
-          })
-        : performQuery({
-              type,
-              payload,
-              resource,
-              action,
-              rest,
-              onSuccess,
-              onFailure,
-              dataProvider,
-              dispatch,
-              logoutIfAccessDenied,
-              allArguments,
-          });
+};
+
+/**
+ * In optimistic mode, the hook dispatches an optimistic action and executes
+ * the success side effects right away. Then it immediately calls the dataProvider.
+ *
+ * We call that "optimistic" because the hook returns a resolved Promise
+ * immediately (although it has an empty value). That only works if the
+ * caller reads the result from the Redux store, not from the Promise.
+ */
+const performOptimisticQuery = ({
```

Comment:
There is a lot of duplication between `performOptimisticQuery` and `performUndoableQuery`. Maybe we can have smaller functions to compose both behaviours?

Reply:
If you look at the details, there are subtle differences here and there (e.g. handling undo and the beforeunload event listener in the performUndoableQuery). I'll extract the functions to make the comparison easier.